upsertGitHubTag/deployment/index.js

```diff
@@ -26,16 +26,17 @@ const verifyGitHub = (req, payload) => {
 };
 
 // Makes a POST request to the given path
-function postEndpoint(path, postBody, callback) {
+function postEndpoint(path, postBody, deliveryId, callback) {
   console.log("POST " + path);
-  console.log(qs.stringify(postBody));
+  console.log(postBody);
 
   const options = url.parse(path);
   options.method = "POST";
   options.headers = {
-    "Content-Type": "application/x-www-form-urlencoded",
+    "Content-Type": "application/json",
```

Comment:
Good catch, yeah it would result in a 415 that would not be retried.

> If we feel it's worth the effort, you could check for a 415 status code in line 59/60 and covert it to a 5xx, which I believe will force a retry later.

Think the fix would be a bit more involved than this. The following table specifies what would happen if the lambda sends the request to a specific webservice version for the old and new lambda.

|             | 1.14.x webservice                                                 | 1.15 webservice    |
|-------------|-------------------------------------------------------------------|--------------------|
| **1.14.x lambda** | OK                                                                | 415 error - Fix: ? |
| **1.15 lambda** | 415 error - Fix by converting 415 to 500 in lambda to retry later | OK                 |

I think your suggestion would be the fix on the second row of the table. We could still get a 415 in the first row scenario. A fix could be to create a hotfix lambda that converts a 415 to 500 in anticipation of the 1.15 release to prevent loss of GitHub App events during the 1.15 deploy.

Is it worth it? I'm not sure. Alternative is to have some down time by bringing down the old webservice first, but I'm not sure where we stand on having down time

Reply:
In my initial comment, I was thinking the lambda would be deployed with the core stack, and so we had a window between the core stack update and the the new Dockstore deploy when we would miss events. But this lambda is deployed with the Dockstore stack.

So what will happen is, I think:

1. Dockstore 1.15 deploy started
2. Deploy adds 1.15 upsert lambda added as a trigger to SQS
3. GitHub events trigger both 1.14 and 1.15 lambdas
4. Both lambdas invoke 1.14 web service; one invocation will succeed, one will fail.
5. 1.15 becomes active web service
6. Both lambdas now invoke 1.15 web service; one invocation fails, one succeeds.
7. We usually leave 1.14 up, if not active, but presumably the 1.14 lambda will continue to invoke the 1.15 web service unless we remove the trigger.

Three questions:

1. Is the above correct?
2. If yes, what is the behavior of events in SQS if half the triggers are succeeding and half are failing?
3. What if instead of bringing down the 1.14 web service in advance, we only remove the 1.14 SQS lambda trigger? I don't know what will happen to the events in that case - hopefully they'll stick around.

>  but I'm not sure where we stand on having down time

I think we're OK with downtime. But I'd prefer it if it doesn't entail missing GitHub events -- I wouldn't want to miss a major release/tag of a major workflow.

Reply:
> 5. 1.15 becomes active web service  
> 6. Both lambdas now invoke 1.15 web service; one invocation fails, one succeeds.

There's a small period of time between 5 and 6 where both 1.14 and 1.15 will be running because it may take a few minutes to set the desired count of the 1.14 ECS service to 0. During this time, both lambdas will invoke dockstore.org and the load balancer will send the request to either the 1.14 or 1.15 webservice, but we don't know which it will send it to. This is the window where we might lose events. If both lambdas send the request to the same webservice, then one out of the two will succeed so no problem. However, if the 1.14 lambda sends the request to the 1.15 webservice and the 1.15 lambda sends the request to the 1.14 webservice, then both will fail and we lose the event. 
> 1. Is the above correct?

Sounds about right, with the note I added above.

> 2. If yes, what is the behavior of events in SQS if half the triggers are succeeding and half are failing?

I don't think we need to worry on the SQS side. The lambda has retries if the status is >= 500. Half will be failing because of a 415 error, so those will not retry unless we add special handling to convert it to a 500. In this case, seems like it's fine to let the 415 fail because the other half will succeed, i.e. the event will be successfully processed by one of the two lambdas.

> 3. What if instead of bringing down the 1.14 web service in advance, we only remove the 1.14 SQS lambda trigger? I don't know what will happen to the events in that case - hopefully they'll stick around.

Bringing down the 1.14 SQS lambda trigger will ensure that only the 1.15 lambda is triggered, but I think it's still possible for the 1.15 lambda to call the 1.14 webservice during the window I initially mentioned where both the 1.14 and 1.15 webservices are running and we could lose the event.

> I think we're OK with downtime. But I'd prefer it if it doesn't entail missing GitHub events -- I wouldn't want to miss a major release/tag of a major workflow.

Hmm, thinking it through, I'm not sure downtime would fix our problem and we could lose events.

1. We set the desired task count of the 1.14 ECS service to 0 to stop the 1.14 webservice. This results in down time and I think visitors to the site receive a 503 code. If we receive events during this down time, the 1.14 lambda tries to invoke the webservice, but it receives a 503. This should signal a retry according to [this](https://github.com/dockstore/lambda/blob/e384ae9430903714c52164a29aef17fd10d00ae4/upsertGitHubTag/deployment/index.js#L287-L293) 
2. We deploy 1.15 Dockstore and the 1.15 webservice runs. We keep the 1.14 Dockstore stack around so that the 1.14 lambda can retry
3. The 1.14 lambda eventually retries and sends the request to the 1.15 webservice which fails with a 415 signalling no retry...If we had a 1.14.x lambda that converts a 415 to 500 then it would retry twice then eventually send the event to its dead queue, but we don't really do anything with that

Looks like we don't have a perfect solution yet for how to prevent loss of GitHub events so I'm gonna do some thinking. I'm wondering if https://ucsc-cgl.atlassian.net/browse/SEAB-5604 can help us by allowing us to record events and retry missed events after a deploy

Reply:
Hmm sounds promising, I'll test it. My theory is that the events should stay in the queue until a trigger is added again. The SQS message retention period is 14 Days and [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html) say that the "lambda polls the queue" as opposed to the queue pushing an event to the lambda

Reply:
Tested in QA by:
1. Removing the lambda trigger from the SQS queue in the SQS console
2. Making a change in my GitHub App workflow repo
3. Adding the trigger back to the SQS queue
4. Checking on Dockstore that my workflow was successfully updated with the event that occurred when the queue didn't have a lambda trigger

So yes seems like the events stay in the queue until a trigger is added again